python/ray/spark/__init__.py

```diff
@@ -0,0 +1,313 @@
+import os
+import shutil
+import subprocess
+import sys
+import time
+import threading
+import logging
+from packaging.version import Version
+
+from .utils import (
+    exec_cmd,
+    check_port_open,
+    get_safe_port,
+    get_spark_session,
+    get_spark_driver_hostname,
+    is_in_databricks_runtime,
+    get_spark_task_assigned_physical_gpus,
+    get_avail_mem_per_ray_worker,
+    get_dbutils,
+)
+
+if not sys.platform.startswith("linux"):
+    raise RuntimeError("Ray on spark ony supports linux system.")
+
+_spark_dependency_error = "ray.spark module requires pyspark >= 3.3"
+try:
+    import pyspark
+    if Version(pyspark.__version__) < Version("3.3"):
+        raise RuntimeError(_spark_dependency_error)
+except ImportError:
+    raise RuntimeError(_spark_dependency_error)
+
+
+def wait_ray_node_available(hostname, port, timeout, error_on_failure):
+    # Wait Ray head node spin up.
+    for _ in range(timeout):
+        time.sleep(1)
+        if check_port_open(hostname, port):
+            break
+
+    if not check_port_open(hostname, port):
+        raise RuntimeError(error_on_failure)
+
+
+class RayClusterOnSpark:
+    """
+    The class is the type of instance returned by `init_cluster` API.
+    It can be used to shutdown the cluster.
+    """
+
+    def __init__(self, address, head_proc, spark_job_group_id, ray_context, ray_head_temp_dir):
+        self.address = address
+        self.head_proc = head_proc
+        self.spark_job_group_id = spark_job_group_id
+        self.ray_context = ray_context
+        self.ray_head_temp_dir = ray_head_temp_dir
+
+    def _cancel_background_spark_job(self):
+        get_spark_session().sparkContext.cancelJobGroup(self.spark_job_group_id)
+
+    def shutdown(self):
+        """
+        Shutdown the ray cluster created by `init_cluster` API.
+        """
+        self.ray_context.disconnect()
+        self._cancel_background_spark_job()
+        self.head_proc.kill()
+        shutil.rmtree(self.ray_head_temp_dir, ignore_errors=True)
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.shutdown()
+
+
+def _convert_ray_node_options(options):
+    return [f"--{k.replace('_', '-')}={str(v)}" for k, v in options.items()]
+
+
+def init_cluster(
+        num_spark_tasks,
+        head_options=None,
+        worker_options=None,
+        ray_temp_dir="/tmp/ray/temp",
```

Comment:
Provide a param to control the `ray_temp_dir` path, this is useful when the default temp dir disk capacity is not sufficient.